### PR TITLE
NO-TICKET - striga. get working with the new home

### PR DESCRIPTION
**Pull request by eduardmaximovich**
## Description of Work
 
Fixed striga at the new home screens (didn't remove old logic)

### CODE REVIEW

File: app/src/main/java/org/p2p/wallet/common/adapter/CommonAnyCellAdapter.kt [hunk 1]
```diff
@@ -9,7 +9,30 @@ import org.p2p.uikit.model.AnyCellItem
...
+        if (index != -1) {
+            val old = oldItems[index] as T
+            val new = transform(old)
+            items = oldItems.apply { set(index, new) }
+            if (animateChanges) {
+                notifyItemChanged(index)
+            } else {
+                // workaround for disabling animation
+                notifyItemChanged(index, Unit)
+            }
```
Comment (1sakov) [inline app/src/main/java/org/p2p/wallet/common/adapter/CommonAnyCellAdapter.kt:32]: looks complex (

**Review by 1sakov — APPROVED** [review]

File: app/src/main/java/org/p2p/wallet/common/adapter/CommonAnyCellAdapter.kt (hunk 1 above)
Comment (eduardmaximovich) [reply app/src/main/java/org/p2p/wallet/common/adapter/CommonAnyCellAdapter.kt:32]: it is, but this way we have ability to change only specific cell without remapping the entire state

File: app/src/main/java/org/p2p/wallet/home/ui/main/delegates/banner/HomeScreenBannerDelegate.kt [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+ *     }
+ * }
+ * ```
+ */
+inline fun <reified T : HomeScreenBanner> homeScreenBannerDelegate(
+    noinline onBindListener: ((binding: ItemBinding, item: T) -> Unit)? = null,
+): AdapterDelegate<List<AnyCellItem>> =
+    adapterDelegateViewBinding<T, AnyCellItem, ItemBinding>(
+        viewBinding = { inflater: LayoutInflater, parent: ViewGroup ->
+            inflater.inflateViewBinding(root = parent, attachToRoot = false)
```
Comment (gslevinkov) [inline app/src/main/java/org/p2p/wallet/home/ui/main/delegates/banner/HomeScreenBannerDelegate.kt:36]: parent.inflateViewBinding(attachToRoot = false)

File: app/src/main/java/org/p2p/wallet/home/ui/main/delegates/banner/HomeScreenBannerDelegate.kt [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+        }
+    ) {
+        @Suppress("UNCHECKED_CAST")
+        bind {
+            when (item) {
+                is StrigaBanner -> (this as AdapterDelegateViewBindingViewHolder<StrigaBanner, ItemBinding>).onBind()
```
Comment (gslevinkov) [inline app/src/main/java/org/p2p/wallet/home/ui/main/delegates/banner/HomeScreenBannerDelegate.kt:42]: holy shit...

File: app/src/main/java/org/p2p/wallet/home/ui/wallet/WalletPresenter.kt [hunk 4]
```diff
@@ -46,6 +68,41 @@ class WalletPresenter(
...
+        launch {
+            strigaUserInteractor.getUserStatusBannerFlow()
+                .collect { banner ->
+                    viewStateFlow.emit(
+                        viewStateFlow.value.copy(
+                            strigaBanner = banner
+                        )
+                    )
+                }
+        }
```
Comment (gslevinkov) [inline app/src/main/java/org/p2p/wallet/home/ui/wallet/WalletPresenter.kt:90]: strigaUserInteractor.getUserStatusBannerFlow()
            .map { viewStateFlow.value.copy(strigaBanner = it) }
            .collect { viewStateFlow.emit(it) }

File: app/src/main/java/org/p2p/wallet/home/ui/wallet/WalletPresenter.kt [hunk 5]
```diff
@@ -46,6 +68,41 @@ class WalletPresenter(
...
+    }
+
+    private fun observeViewState() {
+        launch {
+            viewStateFlow.collect {
+                val items = walletMapper.buildCellItems {
+                    // order is matter
```
Comment (gslevinkov) [inline app/src/main/java/org/p2p/wallet/home/ui/wallet/WalletPresenter.kt:97]:
```suggestion
                    // order matters
```

File: app/src/main/java/org/p2p/wallet/home/ui/wallet/WalletPresenter.kt [hunk 6]
```diff
@@ -109,13 +166,67 @@ class WalletPresenter(
...
+                            view?.showStrigaBannerProgress(isLoading = true)
+
+                            homeInteractor.loadDetailsForStrigaAccounts()
+                                .onSuccess { view?.navigateToStrigaByBanner(statusFromKycBanner) }
+                                .onFailure { view?.showUiKitSnackBar(messageResId = R.string.error_general_message) }
+
+                            view?.showStrigaBannerProgress(isLoading = false)
+                        } else {
+                            view?.navigateToStrigaByBanner(statusFromKycBanner)
+                        }
```
Comment (gslevinkov) [inline app/src/main/java/org/p2p/wallet/home/ui/wallet/WalletPresenter.kt:220]:
```suggestion
                        if (statusFromKycBanner == StrigaKycStatusBanner.VERIFICATION_DONE) {
                            view?.showStrigaBannerProgress(isLoading = true)

                            homeInteractor.loadDetailsForStrigaAccounts()
                                .onSuccess { view?.navigateToStrigaByBanner(statusFromKycBanner) }
                                .onFailure { view?.showUiKitSnackBar(messageResId = R.string.error_general_message) }

                            view?.showStrigaBannerProgress(isLoading = false)
                        } else {
                            view?.navigateToStrigaByBanner(statusFromKycBanner)
                        }
```

let's extract this branch to a method

File: app/src/main/java/org/p2p/wallet/home/ui/wallet/WalletPresenterMapper.kt [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+
+    inner class Builder internal constructor() {
+        private val cellItems = mutableListOf<AnyCellItem>()
+
+        fun mapStrigaKycBanner(banner: StrigaKycStatusBanner?): Builder {
+            banner?.let {
+                cellItems += StrigaBanner(isLoading = false, status = banner)
+            }
+            return this
+        }
```
Comment (gslevinkov) [inline app/src/main/java/org/p2p/wallet/home/ui/wallet/WalletPresenterMapper.kt:24]:
```suggestion
        fun mapStrigaKycBanner(banner: StrigaKycStatusBanner?): Builder = apply {
            banner?.let {
                cellItems += StrigaBanner(isLoading = false, status = banner)
            }
        }
```

**Review by gslevinkov — APPROVED** [review]

File: app/src/main/java/org/p2p/wallet/home/ui/main/delegates/banner/HomeScreenBannerDelegate.kt (hunk 2 above)
Comment (eduardmaximovich) [reply app/src/main/java/org/p2p/wallet/home/ui/main/delegates/banner/HomeScreenBannerDelegate.kt:36]: we already have the created LayoutInflater in the `inflater` variable, creating new one is overhead just to save 17 symbols

File: app/src/main/java/org/p2p/wallet/home/ui/main/delegates/banner/HomeScreenBannerDelegate.kt (hunk 3 above)
Comment (eduardmaximovich) [reply app/src/main/java/org/p2p/wallet/home/ui/main/delegates/banner/HomeScreenBannerDelegate.kt:42]: I could make the `onBind` as not an extensions function, but it must be public (due to inline function requirements), it means that everyone who uses StrigaBanner will have ability to call this function. Thus by reducing scope we disallow the function usage out of the delegate scope

**Review by kabramovp2p — APPROVED** [review]

File: app/src/main/java/org/p2p/wallet/home/ui/wallet/WalletPresenter.kt (hunk 6 above)
Comment (eduardmaximovich) [reply app/src/main/java/org/p2p/wallet/home/ui/wallet/WalletPresenter.kt]: done; extracted to a handler

**Pull request merged**